### DISCUSSION TITLE
Could we emit an event with own params holding, instead of string?

### DISCUSSION THREAD

**Discussion by mikield** [livewire/livewire#4580] [3. Feature ideas]
https://github.com/livewire/livewire/blob/master/src/ComponentConcerns/ReceivesEvents.php#L18

At the given line we can only emit events that would be wrapped into the `Event` object.

Could we see a feature that allowed us to `emit(new Event($params)`?

Sorry for my English, I hope you understand what I am asking for.

In my system, I just want to standardize my events and work with Events the same way as Laravel Event system works.

**Comment by mikield** [comment]
Bump!

**Comment by joshhanley** [comment]
@mikield I can see how that might be useful, but to me Laravel events are different to Livewire events, both in how they run and what they do (Livewire's are javascript events), so to me using a Laravel event object could be confusing for other devs as they're not the same.

What do you think?

**Comment by mikield** [comment]
I did not mean to use Laravel's events. I mean allow passing an Event class.

Let me show a code example :)

```php
    public function emit($event, ...$params)
    {
        if(!$event instanceof Event){
            $event = new Event($event, $params);
        }
        return $this->eventQueue[] = $event;
    }
```

I just wanted to replace my Strings with custom classes, using my folder structure, all events that a component might emit - will be classes that will be placed near the component file.

**Comment by joshhanley** [comment]
@mikield ah right, yeah that could be handy! Maybe it'd need to be called `LivewireEvent` or something like that so it's really clear it's different to a Laravel event. But yeah I like the idea 🙂

You'd need to ensure it serialised to a string as the events are fired using javascript, and there would need to be a way to convert from string back to event object for the listener to use.

PR's are welcome!

**Reply by mikield** [reply]
I have made the changes that resolve my initial thought.
Link to PR: https://github.com/livewire/livewire/pull/4668

If you were thinking that in some other way of usage - please let me know.

Also, if the test that I have created is not enough - please give me some test scenarios to make and I will create tests for them.

**Reply by joshhanley** [reply]
@mikield thanks! Have comment on the PR 🙂
